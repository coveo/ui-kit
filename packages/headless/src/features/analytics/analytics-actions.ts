import {createAsyncThunk, AsyncThunkAction} from '@reduxjs/toolkit';
import {configureAnalytics} from '../../api/analytics/analytics';
import {SearchPageEvents} from 'coveo.analytics/dist/definitions/searchPage/searchPageEvents';
import {SearchAppState} from '../../state/search-app-state';
import {validatePayloadSchema} from '../../utils/validate-payload';
import {StringValue, RecordValue} from '@coveo/bueno';

export const searchPageState = (getState: () => unknown) =>
  getState() as SearchAppState;

export enum AnalyticsType {
  Search,
  Custom,
  Click,
}

export type SearchAction = AsyncThunkAction<
  {analyticsType: AnalyticsType.Search},
  void | {},
  {}
>;

export type CustomAction = AsyncThunkAction<
  {analyticsType: AnalyticsType.Custom},
  {},
  {}
>;

export type ClickAction = AsyncThunkAction<
  {analyticsType: AnalyticsType.Click},
  {},
  {}
>;

export const makeSearchActionType = () => ({
  analyticsType: AnalyticsType.Search as AnalyticsType.Search,
});

export const makeClickActionType = () => ({
  analyticsType: AnalyticsType.Click as AnalyticsType.Click,
});

export interface GenericSearchEventPayload<T = unknown> {
  /** The identifier of the search action (e.g., `interfaceLoad`). */
  evt: SearchPageEvents | string;
  /** The event metadata. */
  meta?: Record<string, T>;
}

/**
 * Logs a generic search event.
 * @param p (GenericSearchEventPayload) The search event payload.
 */
export const logGenericSearchEvent = createAsyncThunk(
  'analytics/generic/search',
  async (p: GenericSearchEventPayload, {getState}) => {
    validatePayloadSchema(p, {
      evt: new StringValue({required: true, emptyAllowed: false}),
      meta: new RecordValue(),
    });
    const {evt, meta} = p;
    const state = searchPageState(getState);
    await configureAnalytics(state).logSearchEvent(
      evt as SearchPageEvents,
      meta
    );
    return makeSearchActionType();
  }
);

/**
 * Logs an interface load event.
 */
export const logInterfaceLoad = createAsyncThunk(
  'analytics/interface/load',
  async (_, {getState}) => {
    const state = searchPageState(getState);
    await configureAnalytics(state).logInterfaceLoad();
    return makeSearchActionType();
  }
);

/**
 * Logs an interface change event.
 */
export const logInterfaceChange = createAsyncThunk(
  'analytics/interface/change',
  async (_, {getState}) => {
    const state = searchPageState(getState);
    await configureAnalytics(state).logInterfaceChange({
      interfaceChangeTo: state.advancedSearchQueries.cq,
    });
    return makeSearchActionType();
  }
);
