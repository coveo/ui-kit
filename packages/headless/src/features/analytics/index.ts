export * from './search-analytics-actions-loader';
export * from './click-analytics-actions-loader';
export * from './generic-analytics-actions-loader';
