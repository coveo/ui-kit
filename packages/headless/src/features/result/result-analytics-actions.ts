import {
  partialDocumentInformation,
  documentIdentifier,
  validateResultPayload,
  makeAnalyticsAction,
  AnalyticsType,
} from '../analytics/analytics-utils';
import {Result} from '../../api/search/search/result';

export const logDocumentOpenThunk = (result: Result) =>
  makeAnalyticsAction(
    'analytics/result/open',
    AnalyticsType.Click,
    (client, state) => {
      validateResultPayload(result);
      return client.logDocumentOpen(
        partialDocumentInformation(result, state),
        documentIdentifier(result)
      );
    }
  );

/**
 * Logs a click event with an `actionCause` value of `documentOpen`.
 * @param result (Result) The result that was opened.
 */
export const logDocumentOpen = (result: Result) =>
  logDocumentOpenThunk(result)();
