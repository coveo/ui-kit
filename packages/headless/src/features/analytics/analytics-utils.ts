import {Result} from '../../api/search/search/result';
import {
  PartialDocumentInformation,
  DocumentIdentifier,
} from 'coveo.analytics/dist/definitions/searchPage/searchPageEvents';
import {SearchAppState} from '../../state/search-app-state';
import {getPipelineInitialState} from '../pipeline/pipeline-state';
import {
  isNullOrUndefined,
  RecordValue,
  Schema,
  StringValue,
} from '@coveo/bueno';
import {Raw} from '../../api/search/search/raw';
import {
  AnalyticsProvider,
  configureAnalytics,
  StateNeededByAnalyticsProvider,
} from '../../api/analytics/analytics';
import {CoveoSearchPageClient, SearchPageClientProvider} from 'coveo.analytics';
import {SearchEventResponse} from 'coveo.analytics/dist/definitions/events';
import {AsyncThunkAction, createAsyncThunk} from '@reduxjs/toolkit';
import {ThunkExtraArguments} from '../../app/store';

export enum AnalyticsType {
  Search,
  Custom,
  Click,
}

export type SearchAction = AsyncThunkAction<
  {analyticsType: AnalyticsType.Search},
  void | {},
  AsyncThunkAnalyticsOptions<StateNeededByAnalyticsProvider>
>;

export type CustomAction = AsyncThunkAction<
  {analyticsType: AnalyticsType.Custom},
  {},
  {}
>;

export type ClickAction = AsyncThunkAction<
  {analyticsType: AnalyticsType.Click},
  {},
  {}
>;

const searchPageState = (getState: () => unknown) =>
  getState() as SearchAppState;

export interface AsyncThunkAnalyticsOptions<
  T extends Partial<StateNeededByAnalyticsProvider>
> {
  state: T;
  extra: ThunkExtraArguments;
}

export const makeAnalyticsAction = <T extends AnalyticsType>(
  prefix: string,
  analyticsType: T,
  log: (
    client: CoveoSearchPageClient,
    state: Partial<SearchAppState>
  ) => Promise<void | SearchEventResponse> | void,
  provider: (state: Partial<SearchAppState>) => SearchPageClientProvider = (
    s
  ) => new AnalyticsProvider(s as StateNeededByAnalyticsProvider)
) => {
  return createAsyncThunk<
    {analyticsType: T},
    void,
    AsyncThunkAnalyticsOptions<StateNeededByAnalyticsProvider>
  >(
    prefix,
    async (
      _,
      {getState, extra: {analyticsClientMiddleware, preprocessRequest, logger}}
    ) => {
      const state = searchPageState(getState);
      const client = configureAnalytics({
        state,
        logger,
        analyticsClientMiddleware,
        preprocessRequest,
        provider: provider(state),
      });
      const response = await log(client, state);
      logger.info(
        {client: client.coveoAnalyticsClient, response},
        'Analytics response'
      );
      return {analyticsType};
    }
  );
};

export const makeNoopAnalyticsAction = <T extends AnalyticsType>(
  analyticsType: T
) => {
  return createAsyncThunk<
    {analyticsType: T},
    void,
    AsyncThunkAnalyticsOptions<StateNeededByAnalyticsProvider>
  >('analytics/noop', async () => {
    return {analyticsType};
  });
};

export const partialDocumentInformation = (
  result: Result,
  state: Partial<SearchAppState>
): PartialDocumentInformation => {
  const resultIndex =
    state.search?.response.results.findIndex(
      ({uniqueId}) => result.uniqueId === uniqueId
    ) || 0;
  return {
    collectionName: result.raw['collection'] || 'default',
    documentAuthor: getDocumentAuthor(result),
    documentPosition: resultIndex + 1,
    documentTitle: result.title,
    documentUri: result.uri,
    documentUriHash: result.raw['urihash'],
    documentUrl: result.clickUri,
    rankingModifier: result.rankingModifier || '',
    sourceName: result.raw['source'],
    queryPipeline: state.pipeline || getPipelineInitialState(),
  };
};

export const documentIdentifier = (result: Result): DocumentIdentifier => {
  return {
    contentIDKey: '@permanentid',
    contentIDValue: result.raw.permanentid,
  };
};

const requiredNonEmptyString = new StringValue({
  required: true,
  emptyAllowed: false,
});

const rawPartialDefinition = {
  collection: new StringValue(),
  urihash: new StringValue(),
  source: new StringValue(),
  permanentid: new StringValue(),
};

const resultPartialDefinition = {
  uniqueId: requiredNonEmptyString,
  raw: new RecordValue({values: rawPartialDefinition}),
  title: requiredNonEmptyString,
  uri: requiredNonEmptyString,
  clickUri: requiredNonEmptyString,
  rankingModifier: new StringValue({required: false, emptyAllowed: true}),
};

function partialRawPayload(raw: Raw): Partial<Raw> {
  return Object.assign(
    {},
    ...Object.keys(rawPartialDefinition).map((key) => ({[key]: raw[key]}))
  );
}

function partialResultPayload(result: Result): Partial<Result> {
  return Object.assign(
    {},
    ...Object.keys(resultPartialDefinition).map((key) => ({
      [key]: result[key as keyof typeof resultPartialDefinition],
    })),
    {raw: partialRawPayload(result.raw)}
  );
}

function getDocumentAuthor(result: Result) {
  const author = result.raw['author'];
  if (isNullOrUndefined(author)) {
    return 'unknown';
  }

  return Array.isArray(author) ? author.join(';') : `${author}`;
}

export const validateResultPayload = (result: Result) =>
  new Schema(resultPartialDefinition).validate(partialResultPayload(result));
