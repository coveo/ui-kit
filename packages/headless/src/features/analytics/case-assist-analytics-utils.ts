import {CaseAssistClient, SearchEventResponse} from 'coveo.analytics';
import {
  StateNeededByCaseAssistAnalytics,
  configureCaseAssistAnalytics,
} from '../../api/analytics/case-assist-analytics';
import {
  PreparableAnalyticsAction,
  makePreparableAnalyticsAction,
} from './analytics-utils';

export type CaseAssistAction = PreparableAnalyticsAction<
  void,
  StateNeededByCaseAssistAnalytics
>;

export const makeCaseAssistAnalyticsAction = (
  prefix: string,
  log: (
    client: CaseAssistClient,
    state: StateNeededByCaseAssistAnalytics
  ) => Promise<void | SearchEventResponse> | void
): PreparableAnalyticsAction<void, StateNeededByCaseAssistAnalytics> => {
  return makePreparableAnalyticsAction(
    prefix,
    async ({
      getState,
      analyticsClientMiddleware,
      preprocessRequest,
      logger,
    }) => {
      const client = configureCaseAssistAnalytics({
        state: getState(),
        logger,
        analyticsClientMiddleware,
        preprocessRequest,
      });
      return {
        log: async () => {
          const response = await log(client, getState());
          logger.info(
            {client: client.coveoAnalyticsClient, response},
            'Analytics response'
          );
        },
      };
    }
  );
};
