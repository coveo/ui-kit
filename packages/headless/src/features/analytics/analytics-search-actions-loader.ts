import {AsyncThunkAction} from '@reduxjs/toolkit';
import {StateNeededByAnalyticsProvider} from '../../api/analytics/analytics';
import {Engine} from '../../app/headless-engine';
import {logClearBreadcrumbs} from '../facets/generic/facet-generic-analytics-actions';
import {AnalyticsType, AsyncThunkAnalyticsOptions} from './analytics-utils';
// import {
//   logFacetBreadcrumb,
//   logFacetClearAll,
//   logFacetDeselect,
//   logFacetSearch,
//   logFacetSelect,
//   logFacetShowLess,
//   logFacetShowMore,
//   logFacetUpdateSort,
// } from '../facets/facet-set/facet-set-analytics-actions'

export interface AnalyticsSearchActionCreators {
  logClearBreadcrumbs(): AsyncThunkAction<
    {analyticsType: AnalyticsType.Search},
    void,
    AsyncThunkAnalyticsOptions<StateNeededByAnalyticsProvider>
  >;
}

/**
 * Returns possible analytics search action creators.
 *
 * @param engine - The headless engine.
 * @returns An object holding the action creators.
 */
export function loadSearchHubActions(
  engine: Engine<object>
): AnalyticsSearchActionCreators {
  engine.addReducers({});

  return {
    logClearBreadcrumbs,
  };
}
