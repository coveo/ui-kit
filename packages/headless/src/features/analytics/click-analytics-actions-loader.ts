import {AsyncThunkAction} from '@reduxjs/toolkit';
import {StateNeededByAnalyticsProvider} from '../../api/analytics/analytics';
import {Result} from '../../api/search/search/result';
import {SearchEngine} from '../../app/search-engine/search-engine';
import {logDocumentOpen} from '../result/result-analytics-actions';
import {AnalyticsType, AsyncThunkAnalyticsOptions} from './analytics-utils';

/**
 * The click analytics action creators.
 */
export interface ClickAnalyticsActionCreators {
  /**
   * The event to log when a result is selected.
   *
   * @param result - The selected result.
   * @returns A dispatchable action.
   */
  logDocumentOpen(result: Result): AsyncThunkAction<
    {
      analyticsType: AnalyticsType.Click;
    },
    void,
    AsyncThunkAnalyticsOptions<StateNeededByAnalyticsProvider>
  >;
}

/**
 * Returns possible click analytics action creators.
 *
 * @param engine - The headless engine.
 * @returns An object holding the action creators.
 */
export function loadClickAnalyticsActions(
  engine: SearchEngine
): ClickAnalyticsActionCreators {
  engine.addReducers({});

  return {
    logDocumentOpen,
  };
}
