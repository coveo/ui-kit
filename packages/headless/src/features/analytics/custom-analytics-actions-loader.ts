import {AsyncThunkAction} from '@reduxjs/toolkit';
import {StateNeededByAnalyticsProvider} from '../../api/analytics/analytics';
import {Engine} from '../../app/headless-engine';
import {AnalyticsType, AsyncThunkAnalyticsOptions} from './analytics-utils';
import {logCustomEvent, CustomEventPayload} from './analytics-actions';

export {CustomEventPayload};

/**
 * The custom analytics action creators.
 */
export interface CustomAnalyticsActionCreators {
  /**
   * Creates a custom analytics event.
   *
   * @param payload - The action creator payload.
   * @returns A dispatchable action.
   */
  logCustomEvent(
    payload: CustomEventPayload
  ): AsyncThunkAction<
    {
      analyticsType: AnalyticsType.Custom;
    },
    void,
    AsyncThunkAnalyticsOptions<StateNeededByAnalyticsProvider>
  >;
}

/**
 * Returns possible custom analytics action creators.
 *
 * @param engine - The headless engine.
 * @returns An object holding the action creators.
 */
export function loadCustomAnalyticsActions(
  engine: Engine<object>
): CustomAnalyticsActionCreators {
  engine.addReducers({});

  return {
    logCustomEvent,
  };
}
