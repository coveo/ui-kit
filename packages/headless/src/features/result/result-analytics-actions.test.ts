import {createRelay} from '@coveo/relay';
import {CoveoSearchPageClient} from 'coveo.analytics';
import {
  SearchEngine,
  buildSearchEngine,
  getSampleSearchEngineConfiguration,
} from '../../app/search-engine/search-engine';
import {buildMockNonEmptyResult} from '../../test/mock-result';
import {logDocumentOpen} from './result-analytics-actions';

jest.mock('@coveo/relay');
jest.mock('coveo.analytics');

async function wait() {
  return new Promise((resolve) => setTimeout(resolve, 0));
}

describe('#logDocumentOpen', () => {
  const testResult = buildMockNonEmptyResult();
  let engine: SearchEngine;
  const makeDocumentOpen = jest.fn();
  const emit = jest.fn();

  beforeEach(() => {
    jest.mocked(CoveoSearchPageClient).mockReturnValue({
      makeDocumentOpen,
    } as unknown as CoveoSearchPageClient);
    jest.mocked(createRelay).mockReturnValue({
      emit,
      getMeta: jest.fn(),
      on: jest.fn(),
      off: jest.fn(),
      updateConfig: jest.fn(),
      clearStorage: jest.fn(),
      version: 'foo',
    });
  });

  describe('when analyticsMode is `legacy`', () => {
    beforeEach(() => {
      engine = buildSearchEngine({
        configuration: {
          ...getSampleSearchEngineConfiguration(),
        },
      });
    });

    it('should call coveo.analytics.makeDocumentOpen properly', async () => {
      engine.dispatch(logDocumentOpen(testResult));
      await wait();

      expect(makeDocumentOpen).toHaveBeenCalledTimes(1);
      expect(makeDocumentOpen.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('when analyticsMode is `next`', () => {
    beforeEach(() => {
      engine = buildSearchEngine({
        configuration: {
          ...getSampleSearchEngineConfiguration(),
          analytics: {
            analyticsMode: 'next',
            trackingId: 'alex',
          },
        },
      });
    });

    it('should call relay.emit properly', async () => {
      engine.dispatch(logDocumentOpen(testResult));
      await wait();

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });
  });
});
