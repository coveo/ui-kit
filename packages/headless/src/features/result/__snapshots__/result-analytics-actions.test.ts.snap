// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#logDocumentOpen when analyticsMode is \`legacy\` should call coveo.analytics.makeDocumentOpen properly 1`] = `
[
  {
    "collectionName": "example collectionName",
    "documentAuthor": "unknown",
    "documentPosition": 1,
    "documentTitle": "example documentTitle",
    "documentUri": "example documentUri",
    "documentUriHash": "example documentUriHash",
    "documentUrl": "example documentUrl",
    "queryPipeline": "",
    "rankingModifier": "example rankingModifier",
    "sourceName": "example sourceName",
  },
  {
    "contentIDKey": "permanentid",
    "contentIDValue": "example contentIDValue",
  },
]
`;

exports[`#logDocumentOpen when analyticsMode is \`next\` should call relay.emit properly 1`] = `
[
  "itemClick",
  {
    "itemMetadata": {
      "author": "unknown",
      "title": "example documentTitle",
      "uniqueFieldName": "permanentid",
      "uniqueFieldValue": "example contentIDValue",
      "url": "example documentUrl",
    },
    "position": 1,
    "searchUid": "",
  },
]
`;
